### PR TITLE
CA-390277: Reduce record usage on CLI cross-pool migrations 

### PR DESCRIPTION
**Pull request by psafont**
Using records in cross-pool migration code is dangerous, as the code interacts with potentially newer hosts. This means that fields in the record might be different from what's expected. In particular, adding an enum field can break the deserialization, and removing a field as well.

Instead, introduce and use a new call `get_all_where`: this calls allows the host to return a filtered list of object references.
 
With these changes, only SR records are used. This is done to minimize the number of calls done.

### CODE REVIEW

**Comment by psafont** [comment]
I've verified manually that it works with and without remote-host, and with and without a remote-network, especially using the name label. For the network, the bridge parameter is strange since it needs the reference, not its name (e.g. xenbr0).


```
# xe vm-migrate live=true remote-username=root remote-password=xxx \
   remote-master=uk-14-05.internal host=uk-14-05 \
   remote-network="Pool-wide network associated with eth0" \
   vm=a8d49c18-69af-fbd4-b4ab-51916073ef15 \
   vif:78502fd1-9b80-cad7-6eb3-cbea04bd1eee=296f1a07-bfc7-8038-7108-f8ddb9cbd322
Performing a storage live migration. Your VM's VDIs will be migrated with the VM.
Will migrate to remote host: uk-14-05, using remote network: Pool-wide network associated with eth0. Here is the VDI mapping:
VDI 807f74a0-be10-48e7-b992-df423e6ff71a -> SR bb7073ab-1e0d-40be-72c8-4653539ce2e7
```

File: ocaml/xapi-cli-server/cli_operations.ml [hunk 1]
```diff
@@ -4579,68 +4579,58 @@ let vm_migrate printer rpc session_id params =
...
-            ( net
-            , Client.Network.get_record ~rpc:remote_rpc
-                ~session_id:remote_session ~self:net
+          match List.assoc_opt "remote-network" params with
+          | Some x -> (
+            match remote Client.Network.get_all_where ~expr:(expr x) with
+            | network :: _ ->
+                network
+            | [] ->
+                failwith (Printf.sprintf "Failed to find network: %s" x)
```
Comment (lindig) [inline ocaml/xapi-cli-server/cli_operations.ml:4616]:
```
let fail fmt = Printf.ksprintf failwith fmt
...
fail "Failed to find network: %s" x"
```

Maybe define in helpers.ml as `fail_with` or `failfmt`

File: ocaml/xapi-cli-server/cli_operations.ml [hunk 2]
```diff
@@ -4579,68 +4579,58 @@ let vm_migrate printer rpc session_id params =
...
+              let expr =
+                Printf.sprintf
+                  {|(field "host"="%s") and (field "management"="true")|}
+                  Ref.(string_of host)
+              in
+              let management_pifs = remote Client.PIF.get_all_where ~expr in
+              match management_pifs with
+              | [] ->
+                  let host_uuid = remote Client.Host.get_uuid ~self:host in
+                  failwith
```
Comment (lindig) [inline ocaml/xapi-cli-server/cli_operations.ml:4628]: See above

File: ocaml/xapi-cli-server/cli_operations.ml [hunk 3]
```diff
@@ -4579,68 +4579,58 @@ let vm_migrate printer rpc session_id params =
...
+          match List.assoc_opt "remote-network" params with
+          | Some x -> (
+            match remote Client.Network.get_all_where ~expr:(expr x) with
+            | network :: _ ->
+                network
+            | [] ->
+                failwith (Printf.sprintf "Failed to find network: %s" x)
+          )
+          | None -> (
+              let expr =
```
Comment (lindig) [inline ocaml/xapi-cli-server/cli_operations.ml:4619]: Expressions are strings?

**Review by lindig — APPROVED** [review]

File: ocaml/xapi-cli-server/cli_operations.ml [hunk 4]
```diff
@@ -4674,15 +4678,12 @@ let vm_migrate printer rpc session_id params =
...
-                (fun (_, pbd_rec) ->
-                  ( pbd_rec.API.pBD_SR
-                  , remote Client.SR.get_record ~self:pbd_rec.API.pBD_SR
-                  )
-                )
-                host_pbds
+              remote Client.PBD.get_all_where ~expr
+              |> List.map (fun pbd ->
+                     let sr = remote Client.PBD.get_SR ~self:pbd in
+                     (sr, remote Client.SR.get_record ~self:sr)
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/xapi-cli-server/cli_operations.ml:4685]: It's not completely safe, but not using means risking a large amount of roundtrips from both server in order to do the filtering.

File: ocaml/idl/datamodel_utils.ml [hunk 5]
```diff
@@ -683,6 +683,33 @@ let messages_of_obj (x : obj) document_order : message list =
...
+          }
+        ]
+    ; msg_result= Some (Set (Ref x.name), "references to all matching objects")
+    ; msg_release=
+        {
+          opensource= []
+        ; internal= x.obj_release.internal
+        ; internal_deprecated_since= None
+        }
+    ; msg_allowed_roles= x.obj_implicit_msg_allowed_roles
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/idl/datamodel_utils.ml:708]: Yes, I used get_all_records_where as template.

File: ocaml/idl/ocaml_backend/gen_db_actions.ml [hunk 6]
```diff
@@ -536,6 +536,12 @@ let db_action api : O.Module.t =
               "let expr' = Xapi_database.Db_filter.expr_of_string expr in"
             ; "get_records_where ~" ^ Gen_common.context ^ " ~expr:expr'"
             ]
+      | FromObject GetAllWhere ->
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/idl/ocaml_backend/gen_db_actions.ml:539]: I believe so, I'll remove this comment

File: ocaml/xapi-cli-server/cli_operations.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vincent-lau) [reply ocaml/xapi-cli-server/cli_operations.ml:4685]: So using SR record is unsafe, but there hasn't been problems using them. And we can fix them once there are changes causing them to become real problems, is that right?

**Review by Vincent-lau — APPROVED** [review]

**Comment by Vincent-lau** [comment]
Looks good to me, as long as the use of SR record is ok....

File: ocaml/xapi-cli-server/cli_operations.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/xapi-cli-server/cli_operations.ml:4685]: > And we can fix them once there are changes causing them to become real problems, is that right?

We have 3 options here:
- Roll back breaking changes when the are found
- Create queries for sorting (mimicking `SORT BY`, similar to the current where) and use that. This seems like reinventing SQL queries.
- Make the client robust to adding fields (ignore them), and removing them (allowing client to fail if they needed it)

File: ocaml/xapi-cli-server/cli_operations.ml [hunk 7]
```diff
@@ -1374,9 +1377,9 @@ let message_destroy (_ : printer) rpc session_id params =
     | None, Some uuids ->
         String.split_on_char ',' uuids
     | Some _, Some _ ->
-        fail "Ambiguous arguments; need one of uuid, uuids"
+        failwith "Ambiguous arguments; need one of uuid, uuids"
```
Comment (robhoes) [inline ocaml/xapi-cli-server/cli_operations.ml:1380]: It's confusing that `failwith` is overridden in this module...

**Review by robhoes — APPROVED** [review]
Some minor comments, but otherwise this looks good.

**Pull request merged**